service/history/historyEngine.go

```diff
@@ -1536,6 +1579,42 @@ func (e *historyEngineImpl) SignalWorkflowExecution(signalRequest *h.SignalWorkf
 		})
 }
 
+func (e *historyEngineImpl) DeleteWorkflowExecutionSignal(deleteRequest *h.DeleteWorkflowExecutionSignalRequest) error {
+	domainID, err := getDomainUUID(deleteRequest.DomainUUID)
+	if err != nil {
+		return err
+	}
+	request := deleteRequest.DeleteRequest
+	execution := workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowExecution.WorkflowId,
+		RunId:      request.WorkflowExecution.RunId,
+	}
+	deleteReq := &persistence.DeleteWorkflowExecutionSignalRequestedRequest{
+		DomainID:        domainID,
+		WorkflowID:      *request.WorkflowExecution.WorkflowId,
+		RunID:           *request.WorkflowExecution.RunId,
+		SignalRequestID: *request.RequestId,
+	}
+
+	context, release, err0 := e.historyCache.getOrCreateWorkflowExecution(domainID, execution)
+	if err0 != nil {
+		return err0
+	}
+	defer release()
+
+	msBuilder, err1 := context.loadWorkflowExecution()
+	if err1 != nil {
+		return err1
+	}
+
+	msBuilder.deleteSignalRequested(*request.RequestId)
+
+	// No need to retry this delete, just leave the record on error until workflow execution is deleted as whole
+	context.executionManager.DeleteSignalRequestedID(deleteReq)
```

Comment:
I don't like that we directly update the persistence here.  I think we should stick to regular scheme of making updates to mutable state and accumulate all the changes and then update persistence at the end.

Reply:
Changed. Add deleteSignalRequestedID in mutable state, and delete in update.